README.md
# seqdataloader
Sequence data label generation and ingestion into deep learning models

## Installation
`pip install seqdataloader`

If you plan to modify the code, you can install it in development mode: 
`pip install -e seqdataloader` 

## Quick Start

tasks.tsv is a 4 column tab-delimited file

Column 1 -- User-specified task name

Column 2 -- path to narrowPeak file (optional if Column 3 is specified)

Column 3 -- path to bigwig file (optional if Column 2 is specified)

Column 4 -- bed file containing user-specified regions to label as ambiguous (i.e. "blacklist")

```
genomewide_labels --task_list tasks.tsv \
		  --outf classificationlabels.SummitWithin200bpCenter.tsv.gz \
		  --output_type gzip \ # (one of gzip, bz2, hdf5, pkl) 
		  --chrom_sizes hg38.chrom.sizes \
		  --bin_stride 50 \
		  --left_flank 400 \
		  --right_flank 400 \
		  --bin_size 200 \
		  --task_hreads 10 \
		  --chrom_threads 4 \
		  --allow_ambiguous \
		  --labeling_approach peak_summit_in_bin_classification 
```
And for regression: 
```
genomewide_labels --task_list tasks.tsv \
       --outf regressionlabels.allbins.hg38.hdf5 \
       --output_type hdf5 \
       --chrom_sizes hg38.chrom.sizes \
       --bin_stride 50 \
       --left_flank 400 \
       --right_flank 400 \
       --chrom_threads 24 \
       --task_threads 2 \
       --label_transformer asinh \ one of None, asinh, log10, log, default is asinh 
       --labeling_approach all_genome_bins_regression
```

labeling_approach can be one of:

    "peak_summit_in_bin_classification"

    "peak_percent_overlap_with_bin_classification"

    "peak_summit_in_bin_regression"

    "peak_percent_overlap_with_bin_regression"
    
    "all_genome_bins_regression"
    

## How to run 
Sample datasets are included in the folder `examples/peak_files_from_encode_for_label_comparison` and `examples/bigwig_files_from_encode_for_label_comparison`

### Executing seqdataloader as a script: 
Execute the script:

`examples/genomewide_labels.sh` for examples on how to generate classification and regression labels on sample datasets.
The script generates binary classification labels (1,0,-1 for ambiguous) or continuous regression labels reflective of bigWig coverage in a bin  in bed file format:

http://mitra.stanford.edu/kundaje/seqdataloader/classificationlabels.50PercentOverlap.tsv.gz

http://mitra.stanford.edu/kundaje/seqdataloader/classificationlabels.SummitWithin200bpCenter.tsv.gz

http://mitra.stanford.edu/kundaje/seqdataloader/regressionlabels.50PercentOverlap.tsv.gz

http://mitra.stanford.edu/kundaje/seqdataloader/regressionlabels.SummitWithin200bpCenter.tsv.gz

Corresponding WashU Browser Tracks with optimal narrowPeak and associated bin labels are here:
http://epigenomegateway.wustl.edu/legacy/?genome=hg38&session=GDB2BTMGnB&statusId=1154897038

### calling seqdataloader as a Python function: 
```
from seqdataloader import *
classification_params={
    'task_list':"tasks.tsv",
    'outf':"classificationlabels.SummitWithin200bpCenter.tsv.gz",
    'output_type':'gzip',
    'chrom_sizes':'hg38.chrom.sizes',
    'chroms_to_keep':['chr21'],
    "store_positives_only":True,
    'bin_stride':50,
    'left_flank':400,
    'right_flank':400,
    'bin_size':200,
    'chrom_threads':10,
    'task_threads':4,
    'allow_ambiguous':True,
    'labeling_approach':'peak_summit_in_bin_classification'
    }
genomewide_labels(classification_params)

regression_params={
    'task_list':"tasks.tsv",
    'outf':"regressionlabels.all_genome_bins_regression.hdf5",
    'output_type':'hdf5',
    'chrom_sizes':'hg38.chrom.sizes',
    'store_values_above_thresh': 0,
    'chroms_to_keep':['chr21'],
    'bin_stride':50,
    'left_flank':400,
    'right_flank':400,
    'bin_size':200,
    'chrom_threads':10,
    'task_threads':4,
    'labeling_approach':'all_genome_bins_regression',
    'label_transformer':'log10',
    'label_transfomer_pseudocount':0.001
    }
genomewide_labels(regression_params)
```


## Dependencies

Please make sure the following dependencies are installed on your system to use SeqDataLoader:
* pybedtools
* pyBigWig 
* pandas
* numpy
* multiprocessing

## Regression label transformations 

In regression mode (   "peak_summit_in_bin_regression", "peak_percent_overlap_with_bin_regression", "all_genome_bins_regression"), the generated labels can be transformed in one of several ways. You can use the arguments `label_transformer` and `label_transformer_pseudocount` to specify the desired tranformation. Allowed values are: 

* asinh --  numpy.arcsinh(values) will be computed (this is the default) 
* None -- no label transformation will be performed 
* log10 --  numpy.log10(values + pseudocount) will be computed using a pseudocount specified by `label_transformer_pseudocount` argument. If this argument is not provided,a default pseudocout of 0.001 is used. 
* log -- numpy.log(values + pseudocount) will be computed using a pseudcount as above. 

## A note on file outputs

The code supports several output types: `hdf5`, `gzip`, `pkl`, `bz2`.
Specify your desired output type with the flag `--output_type`. The default setting for this flag is `gzip`
Please note that the large bottleneck in the code is writing the files to disk. `hdf5` has negligible overhead, but using `gzip` or `bz2` may increase runtime. Timining benchmarks are provided in `examples/genomewide_labels.sh`

You may speed up i/o by writing chromosome outputs to separate files in parallel. This is currently only supported for the `gzip` and `bz2` output types, as i/o is less of a bottleneck for `hdf5` and `pkl` output formats. Use the flag `--split_output_by_chrom` to invoke this parallelized saving of chromosomes.

## Documentation and benchmarks

Testing, benchmarks, and documentation can be found in the `docs` folder


CHANGES.txt
v0.1,01/24/2019 -- Initial release.
v0.11,01/30/2019 -- Added flags --chroms_to_keep, --chroms_to_exclude --store_positives_only. Indices stored as strings rather than tuples for more standard bed format output.
v0.111,01/31/2019 -- default value of False used for flag --store_positives_only when genomewide_labels is called from python script rather than command line script. changed name of module name of genomewide_labels to seqdataloader to avoid user confusion when importing the code 
v0.113,02/01/2019 --Updated how CHROM , START, END are stored in data frame to avoid problems with quoting in the output bed files. (v.112 is skipped due to problem uploading to pypi)
v0.114,02/01/2019 --removed redundant code pass for storing positives only
v0.115,02/01/2019 -- handled an edge case discovered by Soumya 
v0.116,02/07/2019 -- more robus saving of data frames to output hdf5 format
v0.117,02/17/2019 -- added minimum required versions for all dependency packages in setup.py
v0.118,02/17/2019 -- format=table for saving to hdf5 changed as optional,non-default
v0.120,02/26/2019 -- functionality to add a bed file with ambiguous regions for each task. saving as format=table in hdf5 is still optional, but now set to default, as it is desired in most cases
v0.121,02/27/2019 -- ambiguous peaks labeled w/ np.nan
v0.122,02/28/2019 -- write each chromosome to output file individually to reduce ram usage
v0.123,03/02/2019 -- ambig_bed default
[missing documentation]
v0.127,09/05/2019 -- (Av Shrikumar) added functionality to be able to load batches from a downsampled negative set, retrieve labels via a lookup table, Coordinate object is now a namedtuple rather than a full object.
v0.128,09/10/2019 -- (Anna Shcherbina) fixing issue with wheel structure in pypi release



seqdataloader/__init__.py
from __future__ import division, print_function, absolute_import
from . import batchproducers
import argparse
from pybedtools import BedTool
import pyBigWig 
import pandas as pd
import numpy as np 
import pdb
import csv
import sys
from .classification_label_protocols import *
from .regression_label_protocols import * 
from multiprocessing import Pool
from multiprocessing.pool import ThreadPool
import gzip 
import os

#Approaches to determining classification labels
#Others can be added here (imported from classification_label_protocols) 
labeling_approaches={
    "peak_summit_in_bin_classification":peak_summit_in_bin_classification,
    "peak_percent_overlap_with_bin_classification":peak_percent_overlap_with_bin_classification,
    "peak_summit_in_bin_regression":peak_summit_in_bin_regression,
    "peak_percent_overlap_with_bin_regression":peak_percent_overlap_with_bin_regression,
    "all_genome_bins_regression":all_genome_bins_regression
    }

def parse_args():
    parser=argparse.ArgumentParser(description="Generate genome-wide labeled bins for a set of narrowPeak task files ")
    parser.add_argument("--task_list",help="this is a tab-separated file with the name of the task in the first column, the path to the corresponding narrowPeak(.gz) file in the second column (optionally), and the path to the corresponding bigWig file in the third column (optionally, for regression)")
    parser.add_argument("--task_list_sep",default='\t')
    parser.add_argument("--outf",help="output filename that labeled bed file will be saved to.")
    parser.add_argument("--output_type",choices=['gzip','hdf5','pkl','bz2'],default='gzip',help="format to save output, one of gzip, hdf5, pkl, bz2")
    parser.add_argument("--split_output_by_chrom",action="store_true",default=False)
    parser.add_argument("--split_output_by_task",action="store_true",default=False,help="creates a separate output file for each task's labels")
    parser.add_argument("--chrom_sizes",help="chromsizes file for the reference genome. First column is chrom name; second column is chrom size")
    parser.add_argument("--chroms_to_keep",nargs="+",default=None,help="list of chromosomes, as defined in the --chrom_sizes file, to include in label generation. All chromosomes will be used if this argument is not provided. This is most useful if generating a train/test/validate split for deep learning models")
    parser.add_argument("--chroms_to_exclude",nargs="+",default=None,help="list of chromosomes, as defined in the --chrom_sizes file, to exclude in label generation. No chromosomes will be excluded if this argument is not provided. This is most useful if generating a train/test/validate split for deep learning models")
    parser.add_argument("--bin_stride",type=int,default=50,help="bin_stride to shift adjacent bins by")
    parser.add_argument("--left_flank",type=int,default=400,help="left flank")
    parser.add_argument("--right_flank",type=int,default=400,help="right flank")
    parser.add_argument("--bin_size",type=int,default=200,help="flank around bin center where peak summit falls in a positive bin")
    parser.add_argument("--threads",type=int,default=4,help="Number of chromosomes to process at once.")
    parser.add_argument("--subthreads",type=int,default=1,help="Number of tasks to process at once for a given chromosome")
    parser.add_argument("--bigwig_stats",choices=['mean','min','max','coverage','std'],default='mean',help="Value to extract from bigwig file")
    parser.add_argument("--overlap_thresh",type=float,default=0.5,help="minimum percent of bin that must overlap a peak for a positive label")
    parser.add_argument("--allow_ambiguous",default=False,action="store_true")
    parser.add_argument("--store_positives_only",default=False,action="store_true")
    parser.add_argument("--store_values_above_thresh",default=None,type=float,help="for the regression case, determine the minimum row value to include in the output data frame (i.,e. remove bins that are 0 for all tasks by setting this to 0") 
    parser.add_argument("--labeling_approach",choices=["peak_summit_in_bin_classification",
                                                       "peak_percent_overlap_with_bin_classification",
                                                       "peak_summit_in_bin_regression",
                                                       "peak_percent_overlap_with_bin_regression",
                                                       "all_genome_bins_regression"])
    if len(sys.argv)==1:
        parser.print_help(sys.stderr)
        sys.exit(1)       
    return parser.parse_args()

def get_labels_one_task(inputs):
    #unravel the inputs 
    task_name=inputs[0]
    task_bed=inputs[1]
    task_bigwig=inputs[2]
    task_ambig=inputs[3]
    chrom=inputs[4]
    first_coord=inputs[5]
    final_coord=inputs[6]
    args=inputs[7]
    #determine the appropriate labeling approach
    return labeling_approaches[args.labeling_approach](task_name,task_bed,task_bigwig,task_ambig,chrom,first_coord,final_coord,args)    
    
def get_chrom_labels(inputs):
    #unravel inputs 
    chrom=inputs[0]
    chrom_size=inputs[1]
    bed_and_bigwig_dict=inputs[2]
    tasks=inputs[3]
    args=inputs[4] 

    #pre-allocate a pandas data frame to store bin labels for the current chromosome. Fill with zeros    
    #determine the index tuple values
    try:
        chroms,all_start_pos,all_end_pos,first_bin_start,final_bin_start=get_indices(chrom,chrom_size,args)
    except:
        return (chrom,None)
    columns=['CHR','START','END']+list(tasks[0])
    num_entries=len(chroms.values)
    chrom_df = pd.DataFrame(0,index=np.arange(num_entries),columns=columns)
    chrom_df['CHR']=chroms.values
    chrom_df['START']=all_start_pos.values
    chrom_df['END']=all_end_pos.values 
    
    print("pre-allocated df for chrom:"+str(chrom)+"with dimensions:"+str(chrom_df.shape))

    #create a thread pool to label bins, each task gets assigned a thread 
    pool_inputs=[] 
    pool=Pool(args.subthreads)
    for task_name in bed_and_bigwig_dict:
        task_bed=bed_and_bigwig_dict[task_name]['bed']
        task_bigwig=bed_and_bigwig_dict[task_name]['bigwig']
        task_ambig=bed_and_bigwig_dict[task_name]['ambig'] 
        pool_inputs.append((task_name,task_bed,task_bigwig,task_ambig,chrom,first_bin_start,final_bin_start,args))
    bin_values=pool.map(get_labels_one_task,pool_inputs)
    pool.close()
    pool.join()
    
    for task_name,task_labels in bin_values:
        if task_labels is None:
            continue
        chrom_df[task_name]=task_labels
    if args.split_output_by_chrom==True:
        assert args.output_type in ["gzip","bz2"]
        index_label=['Chrom','Start','End']
        chrom_df.to_csv(args.outf+"."+chrom,sep='\t',float_format="%.2f",header=True,index=True,index_label=index_label,mode='wb',compression=args.output_type,chunksize=1000000) 
    return (chrom, chrom_df)

def get_bed_and_bigwig_dict(tasks):
    print("creating dictionary of bed files and bigwig files for each task:")
    bed_and_bigwig_dict=dict()
    for index,row in tasks.iterrows():
        task_name=row[0]
        print(task_name) 
        bed_and_bigwig_dict[task_name]=dict()
        
        #get the peak file associated with the task (if provided)
        if pd.isna(row[0]):
            task_bed=None
        else:
            assert os.path.exists(row[1])
            try:
                task_bed=BedTool(row[1])
            except:
                raise Exception("Could not parse:"+str(row[1]))
        bed_and_bigwig_dict[task_name]['bed']=task_bed
        
        #get the BigWig file associated with the task (if provided)
        if pd.isna(row[2]):
            task_bigwig=None
        else:
            assert os.path.exists(row[2])
            try:
                task_bigwig=pyBigWig.open(row[2])
            except:
                raise Exception("Could not parse:"+str(row[2]))
        bed_and_bigwig_dict[task_name]['bigwig']=row[2] #need to parse the pyBigWig inside Pool
        
        #get the ambiguous peaks
        if pd.isna(row[3]):
            ambig_bed=None
        else: 
            assert os.path.exists(row[3])
            try:
                ambig_bed=BedTool(row[3])
            except:
                raise Exception("Could not parse:"+str(row[3]))
        bed_and_bigwig_dict[task_name]['ambig']=ambig_bed
        
    return bed_and_bigwig_dict

def get_indices(chrom,chrom_size,args):
    final_bin_start=((chrom_size-args.right_flank-args.bin_size)//args.bin_stride)*args.bin_stride
    #final_coord=(chrom_size//args.bin_stride)*args.bin_stride
    first_bin_start=args.left_flank
    if final_bin_start<=first_bin_start:
        print("the chromosome"+chrom+" is too short for the specified settings of --left_flank, --right_flank, --bin_size, skipping")
        return None 
    chroms=[]
    start_pos=[]
    end_pos=[]
    for index in range(first_bin_start,final_bin_start+1,args.bin_stride):
        chroms.append(chrom)
        start_pos.append(index-args.left_flank)
        end_pos.append(index+args.bin_size+args.right_flank)
    return pd.Series(chroms),pd.Series(start_pos),pd.Series(end_pos),first_bin_start,final_bin_start 


def write_output(task_names,full_df,args,mode='w',task_split_engaged=False,outf=None):
    '''
    Save genome-wide labels to disk in gzip, hdf5, or pkl format 
    '''
    if (args.split_output_by_task==True) and (task_split_engaged==False) :
        for task in task_names:
            task_df=full_df[['CHR','START','END',task]]
            write_output([task],task_df,args,mode=mode,task_split_engaged=True,outf=task.replace('/','.')+"."+args.outf)
        return
    if outf==None:
        outf=args.outf        
    all_negative_df=None
    if args.store_positives_only==True:
        #find regions with at least one positive entry per task
        all_negative_df=full_df[['CHR','START','END']][(full_df[task_names]<=0).all(1)]
        full_df=full_df[(full_df[task_names]>0).any(1)]
    if args.store_values_above_thresh is not None:
        all_negative_df=full_df[['CHR','START','END']][(full_df[task_names]<=args.store_values_above_thresh).all(1)]
        full_df=full_df[(full_df[task_names]>args.store_values_above_thresh).any(1)]
    if args.output_type=="gzip":
        full_df.to_csv(outf,sep='\t',header=True,index=False,mode=mode+'b',compression='gzip',chunksize=1000000)
        if all_negative_df is not None:
            all_negative_df.to_csv("universal_negatives."+outf,sep='\t',header=True,index=False,mode=mode+'b',compression='gzip',chunksize=1000000)
    elif args.output_type=="bz2":
        full_df.to_csv(outf,sep='\t',header=True,index=False,mode=mode+'b',compression='bz2',chunksize=1000000)
        if all_negative_df is not None:
            all_negative_df.to_csv("universal_negatives."+outf,sep='\t',header=True,index=False,mode=mode+'b',compression='bz2',chunksize=1000000)
    elif args.output_type=="hdf5":
        full_df=full_df.set_index(['CHR','START','END'])
        if mode=='w':
            append=False
        else:
            append=True
        full_df.to_hdf(outf,key="data",mode=mode, append=append, format='table',min_itemsize={'CHR':30})
        if all_negative_df is not None:
            all_negative_df.set_index(['CHR','START','END'])
            all_negative_df.to_hdf("universal_negatives."+outf,key="data",mode=mode, append=append, format='table',min_itemsize={'CHR':30})
    elif args.output_type=="pkl":
        full_df=full_df.set_index(['CHR','START','END'])        
        full_df.to_pickle(outf,compression="gzip")
        if all_negative_df is not None:
            all_negative_df.set_index(['CHR','START','END'])
            all_negative_df.to_pickle("universal_negatives."+outf,compression="gzip")

def args_object_from_args_dict(args_dict):
    #create an argparse.Namespace from the dictionary of inputs
    args_object=argparse.Namespace()
    #set the defaults
    vars(args_object)['split_output_by_chrom']=False
    vars(args_object)['split_output_by_task']=False
    vars(args_object)['chroms_to_keep']=None
    vars(args_object)['chroms_to_exclude']=None
    vars(args_object)['bin_stride']=50
    vars(args_object)['left_flank']=400
    vars(args_object)['right_flank']=400
    vars(args_object)['bin_size']=200
    vars(args_object)['threads']=1
    vars(args_object)['subthreads']=4
    vars(args_object)['overlap_thresh']=0.5
    vars(args_object)['allow_ambiguous']=True
    vars(args_object)['store_positives_only']=False
    vars(args_object)['store_values_above_thresh']=None
    vars(args_object)['output_hdf5_low_mem']=False
    vars(args_object)['task_list_sep']='\t'
    for key in args_dict:
        vars(args_object)[key]=args_dict[key]
    #set any defaults that are unset 
    args=args_object    
    return args 
        
def genomewide_labels(args):
    if type(args)==type({}):
        args=args_object_from_args_dict(args)
        
    #read in the metadata file with:
    #task names in column 1,
    #path to peak file in column 2,
    #path to bigWig file in column 3
    #path to ambiguous peaks in column 4 (bed) 
    tasks=pd.read_csv(args.task_list,sep=args.task_list_sep,header=None)
    bed_and_bigwig_dict=get_bed_and_bigwig_dict(tasks) 

    #col1: chrom name
    #col2: chrom size 
    chrom_sizes=pd.read_csv(args.chrom_sizes,sep='\t',header=None)

    processed_first_chrom=False
    #create a Pool to process chromosomes in parallel
    print("creating chromosome thread pool")
    pool=ThreadPool(args.threads)
    pool_args=[]
    chrom_order=[] 
    for index,row in chrom_sizes.iterrows():
        chrom=row[0]
        
        #determine whether this chromosome should be included in the label file
        if args.chroms_to_keep!=None:
            if chrom not in args.chroms_to_keep:
                continue
        if args.chroms_to_exclude!=None:
            if chrom in args.chroms_to_exclude:
                continue 
        chrom_order.append(chrom) 
        chrom_size=row[1]
        pool_args.append((chrom,chrom_size,bed_and_bigwig_dict,tasks,args))
    print("launching thread pool")
    processed_chrom_outputs=pool.map(get_chrom_labels,pool_args)
    pool.close()
    pool.join()

    #if the user is happy with separate files for each chromosome, these have already been written to disk. We are done 
    if args.split_output_by_chrom==True:
        exit()
    mode='w'
    for chrom, chrom_df in processed_chrom_outputs:
        #write to output file!
        if chrom_df is None:
            continue 
        print("writing output chromosomes:"+str(chrom))
        write_output(tasks[0],chrom_df,args,mode=mode)
        mode='a'
    print("done!")
    
def main():
    args=parse_args()     
    genomewide_labels(args)
    
if __name__=="__main__":
    main()
